File: clients/android/app/src/main/java/local/visual/corepose/SportsMath.kt
package local.visual.corepose

import kotlin.math.pow
import kotlin.math.sqrt

object SportsMath {
    fun mean(values: DoubleArray): Double {
        if (values.isEmpty()) {
            return 0.0
        }
        return values.sum() / values.size
    }

    fun median(values: DoubleArray): Double {
        if (values.isEmpty()) {
            return 0.0
        }
        val sorted = values.sorted()
        val mid = sorted.size / 2
        return if (sorted.size % 2 == 0) {
            (sorted[mid - 1] + sorted[mid]) / 2.0
        } else {
            sorted[mid]
        }
    }

    fun std(values: DoubleArray): Double {
        if (values.isEmpty()) {
            return 0.0
        }
        val mu = mean(values)
        val varPop = values.sumOf { (it - mu).pow(2) } / values.size
        return sqrt(varPop)
    }

    fun percentile(values: DoubleArray, p: Double): Double {
        if (values.isEmpty()) {
            return 0.0
        }
        val sorted = values.sorted()
        if (sorted.size == 1) {
            return sorted[0]
        }
        val idx = (p / 100.0) * (sorted.size - 1)
        val lo = idx.toInt().coerceIn(0, sorted.lastIndex)
        val hi = (lo + 1).coerceAtMost(sorted.lastIndex)
        val frac = idx - lo
        return sorted[lo] * (1.0 - frac) + sorted[hi] * frac
    }

    fun angleDeg(ax: Double, ay: Double, bx: Double, by: Double, cx: Double, cy: Double): Double {
        val bax = ax - bx
        val bay = ay - by
        val bcx = cx - bx
        val bcy = cy - by
        val nba = sqrt(bax * bax + bay * bay)
        val nbc = sqrt(bcx * bcx + bcy * bcy)
        if (nba < 1e-6 || nbc < 1e-6) {
            return Double.NaN
        }
        val cos = ((bax * bcx + bay * bcy) / (nba * nbc)).coerceIn(-1.0, 1.0)
        return Math.toDegrees(Math.acos(cos))
    }

    fun zeroCrossFreq(series: DoubleArray, fps: Double): Double {
        if (series.size < 4) {
            return 0.0
        }
        val sign = DoubleArray(series.size) { i ->
            val s = kotlin.math.sign(series[i])
            if (s == 0.0) 1.0 else s
        }
        var crosses = 0
        for (i in 1 until sign.size) {
            if (sign[i] * sign[i - 1] < 0) {
                crosses += 1
            }
        }
        val dur = series.size / maxOf(fps, 1.0)
        return 0.5 * crosses / maxOf(dur, 1e-3)
    }
}

File: clients/ios/VisualPose/SportsMath.swift
import Foundation

enum SportsMath {
    static func mean(_ values: [Double]) -> Double {
        if values.isEmpty { return 0 }
        return values.reduce(0, +) / Double(values.count)
    }

    static func median(_ values: [Double]) -> Double {
        if values.isEmpty { return 0 }
        let sorted = values.sorted()
        let mid = sorted.count / 2
        if sorted.count % 2 == 0 {
            return (sorted[mid - 1] + sorted[mid]) / 2.0
        }
        return sorted[mid]
    }

    static func std(_ values: [Double]) -> Double {
        if values.isEmpty { return 0 }
        let mu = mean(values)
        let varPop = values.reduce(0.0) { $0 + pow($1 - mu, 2) } / Double(values.count)
        return sqrt(varPop)
    }

    static func percentile(_ values: [Double], _ p: Double) -> Double {
        if values.isEmpty { return 0 }
        let sorted = values.sorted()
        if sorted.count == 1 { return sorted[0] }
        let idx = (p / 100.0) * Double(sorted.count - 1)
        let lo = min(max(Int(idx), 0), sorted.count - 1)
        let hi = min(lo + 1, sorted.count - 1)
        let frac = idx - Double(lo)
        return sorted[lo] * (1.0 - frac) + sorted[hi] * frac
    }

    static func angleDeg(ax: Double, ay: Double, bx: Double, by: Double, cx: Double, cy: Double) -> Double {
        let bax = ax - bx
        let bay = ay - by
        let bcx = cx - bx
        let bcy = cy - by
        let nba = sqrt(bax * bax + bay * bay)
        let nbc = sqrt(bcx * bcx + bcy * bcy)
        if nba < 1e-6 || nbc < 1e-6 { return .nan }
        let cos = min(max((bax * bcx + bay * bcy) / (nba * nbc), -1.0), 1.0)
        return acos(cos) * 180.0 / .pi
    }

    static func zeroCrossFreq(_ series: [Double], fps: Double) -> Double {
        if series.count < 4 { return 0 }
        let sign = series.map { value -> Double in
            let s = value == 0 ? 0.0 : (value > 0 ? 1.0 : -1.0)
            return s == 0 ? 1.0 : s
        }
        var crosses = 0
        for i in 1..<sign.count {
            if sign[i] * sign[i - 1] < 0 {
                crosses += 1
            }
        }
        let dur = Double(series.count) / max(fps, 1.0)
        return 0.5 * Double(crosses) / max(dur, 1e-3)
    }
}
